src/auth0-session/session/stateful-session.ts

```diff
@@ -0,0 +1,115 @@
+import { signing } from '../utils/hkdf';
+
+const debug = createDebug('stateful-session');
+
+export interface SessionStore<Session> {
+  /**
+   * Gets the session from the store given a session ID.
+   */
+  get(sid: string): Promise<SessionPayload<Session> | null | undefined>;
+
+  /**
+   * Upsert a session in the store given a session ID and `SessionData`.
+   */
+  set(sid: string, session: SessionPayload<Session>): Promise<void>;
+
+  /**
+   * Destroys the session with the given session ID.
+   */
+  delete(sid: string): Promise<void>;
+}
+
+export class StatefulSession<
+  Req,
+  Res,
+  Session extends { [key: string]: any } = { [key: string]: any }
+> extends AbstractSession<Req, Res, Session> {
+  private keys?: Uint8Array[];
+  private store: SessionStore<Session>;
+
+  constructor(protected config: Config, protected Cookies: new () => Cookies) {
+    super(config, Cookies);
+    this.store = config.session.store as SessionStore<Session>;
+  }
+
+  private async getKeys(): Promise<Uint8Array[]> {
+    if (!this.keys) {
+      const secret = this.config.secret;
+      const secrets = Array.isArray(secret) ? secret : [secret];
+      this.keys = await Promise.all(secrets.map(signing));
+    }
+    return this.keys;
+  }
+
+  async getSession(req: Req): Promise<SessionPayload<Session> | undefined | null> {
+    const { name: sessionName } = this.config.session;
+    const cookies = new this.Cookies().getAll(req);
+    const keys = await this.getKeys();
+    const sessionId = await getCookieValue(sessionName, cookies[sessionName], keys);
+
+    if (sessionId) {
+      debug('reading session from %s store', sessionId);
+      return this.store.get(sessionId);
+    }
+    return;
+  }
+
+  async setSession(
+    req: Req,
+    res: Res,
+    session: Session,
+    uat: number,
+    iat: number,
+    exp: number,
+    cookieOptions: CookieSerializeOptions,
+    isNewSession: boolean
+  ): Promise<void> {
+    const { name: sessionName, genid } = this.config.session;
+    const cookieSetter = new this.Cookies();
+    const cookies = cookieSetter.getAll(req);
+    const keys = await this.getKeys();
+    let sessionId = await getCookieValue(sessionName, cookies[sessionName], keys);
+
+    // If this is a new session created by a new login we need to remove the old session
+    // from the store and regenerate the session id to prevent session fixation issue.
+    if (sessionId && isNewSession) {
+      debug('regenerating session id %o to prevent session fixation', sessionId);
+      await this.store.delete(sessionId);
+      sessionId = undefined;
+    }
+
+    if (!sessionId) {
+      sessionId = await genid!(req);
+      debug('generated new session id %o', sessionId);
+      const cookieValue = await generateCookieValue(sessionName, sessionId, keys[0]);
+      cookieSetter.set(sessionName, cookieValue, cookieOptions);
+      cookieSetter.commit(res);
+    }
+    debug('set session %o', sessionId);
+    await this.store.set(sessionId, {
```

Comment:
Just checking, if the session storage operation fails and a new cookie was generated previously (L92), will this cookie still be set?

Reply:
Yes, you'd be left with a cookie pointing to a non existing entry in the store - but the SDK handles this scenario by ignoring it and cleaning it up on the next request.